File: frontend/services/api/searchService.ts
import { API_CONFIG } from '../../config/api';
import { apiGet } from './base';
import { Language } from '../../types';

export type Lang = 'kz' | 'ru' | 'en';

const langMap: Record<Language, Lang> = {
  [Language.KK]: 'kz',
  [Language.RU]: 'ru',
  [Language.EN]: 'en',
};

export interface SearchHit {
  term_id: number;
  slug: string;
  title: string;
  short_definition?: string;
  definition?: string;
  category_id: number;
  rank: number;
  views?: number;
}

export interface AutocompleteHit {
  term_id: number;
  slug: string;
  title: string;
}

export interface SearchResponse {
  meta: {
    page: number;
    size: number;
    total: number;
    pages: number;
  };
  items: SearchHit[];
}

export interface SearchParams {
  q: string;
  lang?: Language;
  category_id?: number;
  letter?: string;
  status?: string;
  sort?: 'newest' | 'oldest' | 'popularity' | 'alphabetical';
  page?: number;
  size?: number;
}

export const searchService = {
  search: async (params: SearchParams): Promise<SearchResponse> => {
    const queryParams = new URLSearchParams({
      q: params.q,
      lang: langMap[params.lang || Language.RU],
      ...(params.category_id && { category_id: params.category_id.toString() }),
      ...(params.letter && { letter: params.letter }),
      ...(params.status && { status: params.status }),
      ...(params.sort && { sort: params.sort }),
      ...(params.page && { page: params.page.toString() }),
      ...(params.size && { size: params.size.toString() }),
    });

    return apiGet<SearchResponse>(
      API_CONFIG.getSearchUrl(`/search?${queryParams}`),
      false
    );
  },

  autocomplete: async (
    query: string,
    lang: Language = Language.RU,
    limit: number = 10
  ): Promise<AutocompleteHit[]> => {
    const queryParams = new URLSearchParams({
      q: query,
      lang: langMap[lang],
      limit: limit.toString(),
    });

    return apiGet<AutocompleteHit[]>(
      API_CONFIG.getSearchUrl(`/search/autocomplete?${queryParams}`),
      false
    );
  },
};

File: backend/apps/search_service/app/services/search_service.py
from typing import Optional
from sqlalchemy.orm import Session
from apps.search_service.app.repositories.search_repo import SearchRepository
from apps.search_service.app.repositories.history_repo import HistoryRepository
from apps.search_service.app.schemas.search import SearchHit, SearchRequestParams
from libs.shared.shared.dto.pagination import PageResponse, PageMeta


class SearchService:
    def __init__(self, db: Session):
        self.search_repo = SearchRepository(db)
        self.history_repo = HistoryRepository(db)
        self.db = db
    
    def search(
        self,
        params: SearchRequestParams,
        user_id: Optional[int] = None
    ) -> PageResponse:
        offset = (params.page - 1) * params.size
        
        results = self.search_repo.search(
            query=params.q,
            language=params.lang.value,
            category_id=params.filters.category_id,
            letter=params.filters.letter,
            status=params.filters.status,
            limit=params.size,
            offset=offset
        )
        
        hits = [
            SearchHit(
                term_id=r["term_id"],
                slug=str(r["term_id"]),
                title=r["title"],
                short_definition=r.get("short_definition"),
                category_id=r.get("category_id", 0),
                rank=float(r.get("rank", 0.0))
            )
            for r in results
        ]
        
        total = len(results)
        pages = (total + params.size - 1) // params.size if total > 0 else 0
        
        self.history_repo.create(
            user_id=user_id,
            query=params.q,
            lang=params.lang.value,
            filters=params.filters.model_dump()
        )
        
        return PageResponse(
            meta=PageMeta(
                page=params.page,
                size=params.size,
                total=total,
                pages=pages
            ),
            items=hits
        )
